server/events/project_command_builder_helpers.go

```diff
@@ -0,0 +1,316 @@
+			return nil, errors.Wrapf(err, "building command for dir %q", plan.RepoRelDir)
+		}
+		cmds = append(cmds, cmd)
+	}
+	return cmds, nil
+}
+
+func buildProjectCommandCtx(
+	ctx *CommandContext,
+	cmd models.CommandName,
+	globalCfg valid.GlobalCfg,
+	commentBuilder CommentBuilder,
+	projectName string,
+	commentFlags []string,
+	repoDir string,
+	repoRelDir string,
+	workspace string,
+	verbose bool) (models.ProjectCommandContext, error) {
+
+	projCfgPtr, repoCfgPtr, err := getCfg(ctx, globalCfg, projectName, repoRelDir, workspace, repoDir)
+	if err != nil {
+		return models.ProjectCommandContext{}, err
+	}
+
+	var projCfg valid.MergedProjectCfg
+	if projCfgPtr != nil {
+		// Override any dir/workspace defined on the comment with what was
+		// defined in config. This shouldn't matter since we don't allow comments
+		// with both project name and dir/workspace.
+		repoRelDir = projCfg.RepoRelDir
+		workspace = projCfg.Workspace
+		projCfg = globalCfg.MergeProjectCfg(ctx.Log, ctx.Pull.BaseRepo.ID(), *projCfgPtr, *repoCfgPtr)
+	} else {
+		projCfg = globalCfg.DefaultProjCfg(ctx.Log, ctx.Pull.BaseRepo.ID(), repoRelDir, workspace)
+	}
+
+	if err := validateWorkspaceAllowed(repoCfgPtr, repoRelDir, workspace); err != nil {
+		return models.ProjectCommandContext{}, err
+	}
+
+	automerge := DefaultAutomergeEnabled
+	parallelApply := DefaultParallelApplyEnabled
+	parallelPlan := DefaultParallelPlanEnabled
+	if repoCfgPtr != nil {
+		automerge = repoCfgPtr.Automerge
+		parallelApply = repoCfgPtr.ParallelApply
+		parallelPlan = repoCfgPtr.ParallelPlan
+	}
+
+	return buildCtx(
+		ctx,
+		cmd,
+		commentBuilder,
+		projCfg,
+		commentFlags,
+		automerge,
+		parallelApply,
+		parallelPlan,
+		verbose,
+		repoDir,
+	), nil
+}
+
+// validateWorkspaceAllowed returns an error if repoCfg defines projects in
+// repoRelDir but none of them use workspace. We want this to be an error
+// because if users have gone to the trouble of defining projects in repoRelDir
+// then it's likely that if we're running a command for a workspace that isn't
+// defined then they probably just typed the workspace name wrong.
+func validateWorkspaceAllowed(repoCfg *valid.RepoCfg, repoRelDir string, workspace string) error {
+	if repoCfg == nil {
+		return nil
+	}
+
+	projects := repoCfg.FindProjectsByDir(repoRelDir)
+
+	// If that directory doesn't have any projects configured then we don't
+	// enforce workspace names.
+	if len(projects) == 0 {
+		return nil
+	}
+
+	var configuredSpaces []string
+	for _, p := range projects {
+		if p.Workspace == workspace {
+			return nil
+		}
+		configuredSpaces = append(configuredSpaces, p.Workspace)
+	}
+
+	return fmt.Errorf(
+		"running commands in workspace %q is not allowed because this"+
+			" directory is only configured for the following workspaces: %s",
+		workspace,
+		strings.Join(configuredSpaces, ", "),
+	)
+}
+
+// buildCtx is a helper method that handles constructing the ProjectCommandContext.
+func buildCtx(ctx *CommandContext,
```

Comment:
now that we've ripped this out into another file. Can i just say I despise that we have a function named `buildProjectCommandCtx` and `buildCtx`.  I'd be in favor of just combining these functions, or naming these better + separating out the logic accordingly to reflect the naming.